File: hannoob/python/day_1.py
from toolz import thread_last
import re

def read_lines_from_file(filename):
    with open(filename, 'r') as file:
        lines = file.readlines()
    return lines

def convert_strings_to_numbers(line):
    numbers = {"one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9, "zero": 0}
    
    for number,value in numbers.items():
        line = line.replace(number, f'{number}{value}{number}')

    return line

def calculate_number_from_string(line):
    digits = [c for c in line if c.isdigit()]
    return int(digits[0] + digits[-1])

def calculate_problem1(lines):
    return thread_last(
        lines,
        (map, calculate_number_from_string),
        sum
    )

def calculate_problem2(lines):
    return thread_last(
        lines,
        (map, convert_strings_to_numbers),
        (map, calculate_number_from_string),
        sum
    )

lines = read_lines_from_file('input.txt')
print(f"problem 1: {calculate_problem1(lines)}")
print(f"problem 2: {calculate_problem2(lines)}")

File: deanddm/Day1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AOC2023
{
    internal class Day1
    {
        String[] lines = File.ReadAllLines("input.day1.txt");
        Dictionary<string, string> numbers = new Dictionary<string, string> { { "one", "o1e" }, { "two", "t2o" }, { "three", "t3e" }, { "four", "f4r" }, { "five", "f5v" }, { "six", "s6x" }, { "seven", "s7n" }, { "eight", "e8t" }, { "nine", "n9n" } };
        Func<string, int> numberExtractor = line => Int32.Parse($"{line.FirstOrDefault(x => Char.IsDigit(x))}{line.LastOrDefault(x => Char.IsDigit(x))}");

        internal void ExecutePart1()
        {
            int sum = lines.ToList().Sum(numberExtractor);
            Console.WriteLine("Day 1, Part 1:" + sum);
        }

        internal void ExecutePart2()
        {
            for (int i = 0; i < lines.Count(); i++)
            {
                foreach (var number in numbers)
                    lines[i] = lines[i].Replace(number.Key, number.Value);
            }

            int sum = lines.ToList().Sum(numberExtractor);
            Console.WriteLine("Day 1, Part 2:" + sum);
        }
    }
}
